akka-actor/src/main/scala/akka/actor/ActorCell.scala

```diff
@@ -421,6 +421,9 @@ private[akka] class ActorCell(
     with dungeon.DeathWatch
     with dungeon.FaultHandling {
 
+  private[this] var _props = _initialProps
```

Comment:
@jrudolph  One issue I see here is a potential difference in visibility of this write—both during construction but also after failing—how do other threads observe these writes?

Reply:
Do you have an example how that could be observed? The field backs the public `props` method of `ActorContext` so the only valid access to that would be from the actor context itself, wouldn't it? After creation of the cell, the `Create` system message is sent to actor to actually instantiate the `Actor` instance. The execution of the `Actor` constructor would be the first time the `ActorContext` could be accessed. However, since that's done through a system message send going through volatile writes, it seems the whole ActorCell should be published safely to that context.

Not sure about the exact flow during failure. I'd assume that any failure handling would only be executed during system message handling, so not sure how one would observe a change in that value from the outside.

Incidentally, in this version, the value of props has less significance than before. Before the value here was an indicator of the failure status of the actor which is now tracked explicitly with the new `_failed` variable.

Reply:
@jrudolph For the cases where someone has inadvertently closed over `self`—will they observe the writes?

Reply:
@viktorklang, as far as I can tell, that won't be different than the status-quo. `self` being an ActorRef, users don't have direct access to the `props`. Of course they can use the ref as if that was still alive and get failures, but the value of the `props` isn't relevant in that case. 

But I may be wrong, I don't know the internals well enough. There may be some cases where this can be a problem.

Reply:
btw, when I say status-quo I'm referring to that save value being changed with reflection.

Reply:
@renatocaval I think it's mostly around how JVMs handle writes to finals (if they inject barriers or not).

Reply:
Not clear if you are saying it does or if you are saying it's an open question. 

In any case, we can have another open question. If there are barriers for final fields, are they took into consideration when updating it via reflection? 

I don't have answers for it. Is that a concern given that `self` is an `ActorRed`? Which other cases could that be an issue?

Reply:
@renatocaval In order to answer whether the proposed change is safe or not, one must first check if reflectively writing a final field is program-semantically equivalent to a plain field write or not.

Reply:
> @renatocaval In order to answer whether the proposed change is safe or not, one must first check if reflectively writing a final field is program-semantically equivalent to a plain field write or not.

We only have to do that if we want to keep compatibility with buggy code. If we cannot think of a valid code pattern that would observe the difference, why would we care?

Reply:
@viktorklang 

> @renatocaval I think it's mostly around how JVMs handle writes to finals (if they inject barriers or not).

It's only written twice: once at initialization and once at termination. For initialization, I haven't looked up the semantics but I think they are different because barriers are only guaranteed for final field writes in constructors. But there it doesn't matter because the value can only be observed from the Actor and the whole cell/context is published safely to the the thread where the actor will run.

For termination, a different thread might observe a different value (than `TerminatedProps` to which it is set) for a short time, but the only way to observe it there is if you have shared the context before (which is discouraged for user code, so only internal code would be affected). Even so, it would be a race condition any way, for it to be observable you would have to assume a happens before relationship between stopping an actor and then reading props. I think that would only make sense for internals. I had a quick look at internal callers to `props` but nothing dangerous surfaced.